variant_remapping_tools/reads_to_remapped_variants.py

```diff
@@ -104,33 +103,53 @@ def calculate_new_alleles(old_ref, new_ref, old_alt, is_reverse_strand):
+        current_reads = []
+        for read in inbam:
+            if read.query_name == current_read_name:
+                current_reads.append(read)
+            else:
+                if current_read_name:
+                    yield current_reads
+                current_reads = [read]
+            current_read_name = read.query_name
+        yield current_reads
+
+
+def process_bam_file(bam_file_path, output_file, flank_length, score_perc, diff_AS_XS_perc, max_nb_alignment,
+                     new_genome):
 
     # Calculate the score cutoff based on flanking seq length
     score_cutoff = -(flank_length * score_perc)
     diff_cutoff = flank_length * diff_AS_XS_perc
     counter = Counter()
-
     fasta = pysam.FastaFile(new_genome)
     with open(output_file, 'w') as outfile:
-        for read in bamfile:
-            if is_read_valid(read, counter, flank_length, score_cutoff, diff_cutoff):
-                varpos, perf_counter, local_region_size = calculate_variant_position(read, flank_length)
-                if is_local_alignment_perfect(counter, perf_counter, local_region_size):
-                    name = read.query_name
-                    info = name.split('|')
-                    old_ref = info[2]
-                    old_alt = info[3]
-
-                    new_ref = fetch_bases(fasta, read.reference_name, len(old_ref), varpos)
-                    new_ref, new_alt = calculate_new_alleles(old_ref, new_ref, old_alt, read.is_reverse)
-
-                    outfile.write(
-                        '%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n' % (read.reference_name, varpos, info[4], new_ref, new_alt,
-                                                              info[5], info[6], info[7]))
+        for read_group in group_reads(bam_file_path):
+            if len(read_group) <= max_nb_alignment:
+                counter['total'] += 1
+                for read in read_group:
+                    if is_read_valid(read, counter, flank_length, score_cutoff, diff_cutoff):
+                        varpos, perf_counter, local_region_size = calculate_variant_position(read, flank_length)
+                        if is_local_alignment_perfect(counter, perf_counter, local_region_size):
+                            name = read.query_name
+                            info = name.split('|')
+                            old_ref = info[2]
+                            old_alt = info[3]
+
+                            new_ref = fetch_bases(fasta, read.reference_name, len(old_ref), varpos)
+                            new_ref, new_alt = calculate_new_alleles(old_ref, new_ref, old_alt, read.is_reverse)
+
+                            outfile.write(
+                                '%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n' % (read.reference_name, varpos, info[4], new_ref, new_alt,
+                                                                  info[5], info[6], info[7]))
+            else:
+                counter['Too many alignments'] += 1
```

Comment:
It's probably a matter of personal taste, so you don't need to apply this, but I think it might be better with less indentation:

```suggestion
            if len(read_group) > max_nb_alignment:
                counter['Too many alignments'] += 1
                continue
            counter['total'] += 1
            for read in read_group:
                if not is_read_valid(read, counter, flank_length, score_cutoff, diff_cutoff):
                    continue
                varpos, perf_counter, local_region_size = calculate_variant_position(read, flank_length)
                if not is_local_alignment_perfect(counter, perf_counter, local_region_size):
                    continue

                info = read.query_name.split('|')
                old_ref = info[2]
                old_alt = info[3]
                new_ref = fetch_bases(fasta, read.reference_name, len(old_ref), varpos)
                new_ref, new_alt = calculate_new_alleles(old_ref, new_ref, old_alt, read.is_reverse)

                outfile.write(
                    '%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n' % (read.reference_name, varpos, info[4], new_ref, new_alt,
                                                      info[5], info[6], info[7]))
```

Reply:
This was refactor in subsequent PR so I'll leave it as is but I agree with the suggestion.